dipy/segment/bundles.py

```diff
@@ -0,0 +1,406 @@
+import numpy as np
+from dipy.tracking.streamline import (transform_streamlines,
+                                      set_number_of_points, nbytes,
+                                      select_random_set_of_streamlines)
+from dipy.segment.clustering import qbx_and_merge
+from dipy.tracking.distances import (bundles_distances_mdf,
+                                     bundles_distances_mam)
+from dipy.align.streamlinear import (StreamlineLinearRegistration,
+                                     BundleMinDistanceMetric,
+                                     BundleSumDistanceMatrixMetric,
+                                     BundleMinDistanceAsymmetricMetric)
+from dipy.align.bundlemin import distance_matrix_mdf
+from time import time
+from itertools import chain
+
+from dipy.tracking.streamline import Streamlines
+from nibabel.affines import apply_affine
+
+
+class RecoBundles(object):
+
+    def __init__(self, streamlines, cluster_map=None, clust_thr=15, nb_pts=20,
+                 verbose=True):
+        """ Recognition of bundles
+
+        Extract bundles from a participants' tractograms using model bundles
+        segmented from a different subject or an atlas of bundles.
+        See [Garyfallidis17]_ for the details.
+
+        Parameters
+        ----------
+        streamlines : Streamlines
+            The tractogram in which you want to recognize bundles.
+        cluster_map : QB map
+            Provide existing clustering to start RB faster (default None).
+        clust_thr : float
+            Distance threshold in mm for clustering `streamlines`
+        nb_pts : int
+            Number of points per streamline (default 20)
+
+        Notes
+        -----
+        Make sure that before creating this class that the streamlines and
+        the model bundles are roughly in the same space.
+        Also default thresholds are assumed in RAS 1mm^3 space. You may
+        want to adjust those if your streamlines are not in world coordinates.
+
+        References
+        ----------
+        .. [Garyfallidis17] Garyfallidis et al. Recognition of white matter
+            bundles using local and global streamline-based registration and
+            clustering, Neuroimage, 2017.
+        """
+        self.clust_thr = clust_thr
+        self.streamlines = streamlines
+
+        self.nb_streamlines = len(self.streamlines)
+        self.verbose = verbose
+
+        self.start_thr = [40, 25, 20]
+
+        if cluster_map is None:
+            self._cluster_streamlines(clust_thr=clust_thr, nb_pts=nb_pts)
+        else:
+            if self.verbose:
+                t = time()
+
+            self.cluster_map = cluster_map
+            self.cluster_map.refdata = self.streamlines
+            self.centroids = self.cluster_map.centroids
+            self.nb_centroids = len(self.centroids)
+            self.indices = [cluster.indices for cluster in self.cluster_map]
```

Comment:
Thank you for the update, it's incredibly better. You can go further by removing `self.clust_thr` and `self.indices` (never used), and `self.nb_centroids` (only used in a `print`, so simply use  `len`). I think I won't complain after that :)

Reply:
I am removing self.clust_thr but the others are useful to stay there. I have reduced number of attributes incredibly there is no need to move directly and made _cluster_streamlines a "private" function. So, you should be very happy. Some utility attributes are useful.